From 8d40057188245122b0446077d6fc92dfdce062ae Mon Sep 17 00:00:00 2001
From: Weiguang Kong <weiguang.kong@nxp.com>
Date: Fri, 7 Jul 2017 12:38:09 +0800
Subject: [PATCH 2094/5242] MLK-15934-3: ASoc: fsl: add hifi4 firmware's
 status transfer support

commit  45861465e384c2d969cb2ed65dec36154ed95e5e from
https://source.codeaurora.org/external/imx/linux-imx.git

1. add cases to receive error value from hifi4 firmware and
   return this error to hifi4 driver's caller.
2. add cases to receive input over indicator variable from
   hifi4 dirver's caller and pass this value to hifi4 firmware

Signed-off-by: Weiguang Kong <weiguang.kong@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 include/uapi/linux/mxc_hifi4.h |    1 +
 sound/soc/fsl/fsl_hifi4.c      |    3 +++
 sound/soc/fsl/fsl_hifi4.h      |    1 +
 3 files changed, 5 insertions(+)

diff --git a/include/uapi/linux/mxc_hifi4.h b/include/uapi/linux/mxc_hifi4.h
index b1d631b..8e01890 100644
--- a/include/uapi/linux/mxc_hifi4.h
+++ b/include/uapi/linux/mxc_hifi4.h
@@ -44,6 +44,7 @@ struct decode_info {
 	int   out_buf_size;
 	int   out_buf_off;
 	unsigned int cycles;
+	unsigned int input_over;
 };
 
 struct prop_info {
diff --git a/sound/soc/fsl/fsl_hifi4.c b/sound/soc/fsl/fsl_hifi4.c
index c2c8e01..bf8aa65 100644
--- a/sound/soc/fsl/fsl_hifi4.c
+++ b/sound/soc/fsl/fsl_hifi4.c
@@ -621,6 +621,8 @@ static long fsl_hifi4_decode_frame(struct fsl_hifi4 *hifi4_priv,
 	codec_iobuf_info->out_buf_size_max = hifi4_priv->out_buf_size;
 	codec_iobuf_info->out_cur_offset   = 0;
 
+	codec_iobuf_info->input_over   = decode_info.input_over;
+
 	init_completion(&hifi4_priv->cmd_complete);
 	hifi4_priv->is_done = 0;
 
@@ -647,6 +649,7 @@ static long fsl_hifi4_decode_frame(struct fsl_hifi4 *hifi4_priv,
 	decode_info.in_buf_off = codec_iobuf_info->inp_cur_offset;
 	decode_info.out_buf_off = codec_iobuf_info->out_cur_offset;
 	decode_info.cycles = codec_iobuf_info->cycles;
+	decode_info.input_over = codec_iobuf_info->input_over;
 
 	ret = copy_to_user(user, &decode_info, sizeof(decode_info));
 	if (ret) {
diff --git a/sound/soc/fsl/fsl_hifi4.h b/sound/soc/fsl/fsl_hifi4.h
index 3cb8356..e7c9dee 100644
--- a/sound/soc/fsl/fsl_hifi4.h
+++ b/sound/soc/fsl/fsl_hifi4.h
@@ -56,6 +56,7 @@ struct icm_cdc_iobuf_t {
 	u32 out_cur_offset;		/* init by APU, updated by DPU */
 	s32 ret;
 	u32 cycles;		        /* consumed cycles during executing */
+	u32 input_over;			/* indicate external stream is over*/
 };
 
 struct icm_cdc_uinp_t {
-- 
1.7.9.5

